doc/book/app_server/luajit_memprof.rst

```diff
@@ -0,0 +1,527 @@
+    INTERNAL: 1     65536   0
+
+    REALLOCATIONS
+
+    DEALLOCATIONS
+    INTERNAL: 19998 0       558816
+            Overrides:
+                    @format_concat.lua:9, line 10
+    @format_concat.lua:9, line 10: 2 0       98304
+            Overrides:
+                    @format_concat.lua:9, line 10
+
+The reasonable questions regarding the report can be:
+
+*   Why are there no allocations related to the ``concat()`` function?
+*   Why the amount of allocations is not a round number?
+*   Why are there approximately 20K allocations instead of 10K?
+
+First of all, LuaJIT doesn't create a new string if the string with the same
+payload exists (see details on `lua-users.org/wiki <http://lua-users.org/wiki/ImmutableObjects>`_).
+This is called the `string interning <https://en.wikipedia.org/wiki/String_interning>`_.
+So, when the string is
+created via the ``format()`` function, there is no need to create the same
+string via the ``concat()`` function, and LuaJIT just uses the previous one.
+
+That is also the reason why the amount of allocations is not the round number
+as can be expected from the cycle operator ``for i = 1, 10000...``:
+Tarantool creates some
+strings for internal needs and built‑in modules, so some strings already exist.
+
+But why are there so many allocations? It's almost twice as big as the expected
+amount. This is because the ``string.format()`` built‑in function creates
+another string necessary for the ``%s`` identifier, so there are two allocations
+for each iteration: for ``tostring(i)`` and for ``string.format("%sa", string_i_value)``.
+You can see the difference in behaviour by adding the
+``local _ = tostring(i)`` line between lines 22 and 23.
+
+To profile only the ``concat()`` function, comment line 23, namely,
+``local f = format(i)`` and run the profiler.
+
+The profiler's output is the following:
+
+..  code-block:: console
+
+    ALLOCATIONS
+    @format_concat.lua:4, line 5: 10000     284411  0
+
+    REALLOCATIONS
+
+    DEALLOCATIONS
+    INTERNAL: 10000 0       218905
+            Overrides:
+                    @format_concat.lua:4, line 5
+    @format_concat.lua:4, line 5: 1 0       32768
+
+**Q**: But what will change if the JIT compilation is enabled?
+
+**A**: In the :ref:`code <profiler_usage_example03>`, comment line 2, namely,
+``jit.off()`` and run
+the profiler . Now, there are only 56 allocations in the report, and all other
+allocations are JIT-related (see also the related
+`dev issue <https://github.com/tarantool/tarantool/issues/5679>`_):
+
+..  code-block:: console
+
+    ALLOCATIONS
+    @format_concat.lua:4, line 5: 56        1112    0
+    @format_concat.lua:0, line 0: 4 640     0
+    INTERNAL: 2     382     0
+
+    REALLOCATIONS
+
+    DEALLOCATIONS
+    INTERNAL: 58    0       1164
+            Overrides:
+                    @format_concat.lua:4, line 5
+                    INTERNAL
+
+This happens because a trace is compiled after 56 iterations, and the
```

Comment:
Nit: I believe it's worth to mention that *56 is a default value of the one compiler parameter (namely, [hotloop](luajit.org/running.html#opt_O))*.  Hence this particular trace is compiled after 56 iterations. At the same time, I don't want to be super-accurate and pedantic if it can spoil the perception of the document. So, feel free to ignore.

Reply:
I think for developers it's OK to have such details. Otherwise, it's not really clear if it's a random number or is some logic behind that.